src/routes/class-base-route.php

```diff
@@ -79,80 +94,288 @@ protected function get_route_name() : string {
+        'data' => $data,
+      )
+    );
+  }
+
   /**
    * Ensure correct response for rest using error handler function.
    *
    * @param  string $response_key Which response to get.
    * @param  array  $data         (Optional) Data to pass to response handler.
    *
-   * @return \WP_Error|array \WP_Error instance with error message and status or array.
+   * @return \WP_REST_Response|WP_Error|WP_HTTP_Response|mixed
    */
   protected function rest_response_handler( string $response_key, array $data = array() ) {
-    return \rest_ensure_response( $this->defined_responses( $response_key, $data ) );
+    $responses = array_merge( $this->route_responses(), $this->all_responses() );
+
+    $response = $responses[ $response_key ] ?? [
+      'code' => 400,
+      'message' => esc_html__( 'Undefined response', 'eightshift-forms' ),
+    ];
+
+    $response['data'] = $data;
+    return \rest_ensure_response( $response );
   }
 
   /**
    * Define a list of responses for this route.
    *
-   * @param  string $response_key Which response to get.
-   * @param  array  $data         (Optional) Data to pass to response handler.
    * @return array
    */
-  abstract protected function defined_responses( string $response_key, array $data = array() ): array;
+  protected function route_responses(): array {
+    return [];
+  }
+
+  /**
+   * A list of all responses.
+   *
+   * @return array
+   */
+  protected function all_responses(): array {
+    return [
+      'wrong-captcha' => [
+        'code' => 429,
+        'message' => esc_html__( 'Wrong captcha answer.', 'eightshift-forms' ),
+      ],
+      'send-email-error' => [
+        'code' => 400,
+        'message' => esc_html__( 'Error while sending an email.', 'eightshift-forms' ),
+      ],
+      'missing-params' => [
+        'code' => 400,
+        'message' => esc_html__( 'Missing one or more required GET parameters to process the request.', 'eightshift-forms' ),
+      ],
+      'missing-post-params' => [
+        'code' => 400,
+        'message' => esc_html__( 'Missing one or more required POST parameters to process the request.', 'eightshift-forms' ),
+      ],
+      'integration-not-used' => [
+        'code' => 400,
+        'message' => sprintf( esc_html__( 'This form integration is not used, please add a filter returning all necessary info.', 'eightshift-forms' ) ),
+      ],
+      'authorization-invalid' => [
+        'code' => 400,
+        'message' => sprintf( esc_html__( 'Unauthorized request', 'eightshift-forms' ) ),
+      ],
+
+      // Buckaroo specific.
+      'buckaroo-missing-keys' => [
+        'code' => 400,
+        'message' => esc_html__( 'Not all Buckaroo keys are set', 'eightshift-forms' ),
+      ],
+      'buckaroo-request-exception' => [
+        'code' => 400,
+        'message' => esc_html__( 'Error ocurred, unable to redirect to Buckaroo', 'eightshift-forms' ),
+      ],
+    ];
+  }
```

Comment:
This entire class seems a bit overloaded for a template.

The idea of base classes is to be the template on which **all other** classes can be build from and reduce duplication. To me, it looks like this base class does waaaay too much. That is, I could see a rest class that wouldn't use any of these, and that kinda defeats the purpose of the template method pattern.

Reply:
Dunno, all routes in forms use almost all functionality here, like before all this code was in each separate class and there was a lot of code duplication.